specs: include source provenance in `spec.json` and package hash

tgamblin opened the pull request:
We've included a package hash in Spack since #7193 for CI, and we started using it on the spec in #28504. However, what goes into the package hash is opaque.

We want this information to be available *from the concrete spec* so that it can be retrieved as provenance *without* referring to a `package.py` file.  Why?

1. We want to be able to look up authoritative checksums, versions, etc. and pass them through scanners long *after* an installation takes place.
2. Right now you can look at the version on a `Spec`, then look up the checksum for a version from the `package.py`, but that's not reliable in the long term.  We may remove versions from `package.py` (in fact, we might want to as they become less relevant), and checksums for particular versions/releases may change over time.
3. Only the version is stored explicitly on the spec; that's not sufficient to know *exactly* what hashes a spec was built with.

Adding hashes explicitly gets around this and gets us closer to having all the information we need for a complete SBOM.

Here's what `spec.json` looked like before:

```json
{
  "spec": {
    "_meta": {
      "version": 3
    },
    "nodes": [
      {
        "name": "zlib",
        "version": "1.2.12",
        ...
        "patches": [
          "0d38234384870bfd34dfcb738a9083952656f0c766a0f5990b1893076b084b76"
        ],
        "package_hash": "pthf7iophdyonixxeed7gyqiksopxeklzzjbxtjrw7nzlkcqleba====",
        "hash": "ke4alug7ypoxp37jb6namwlxssmws4kp"
      }
    ]
  }
}
```

The `package_hash` there is a hash of the concatenation of:

* A canonical hash of the `package.py` recipe, as implemented in #28156;
* `sha256`'s of patches applied to the spec; and
* Archive `sha256` sums of archives or commits/revisions of repos used to build the spec.

There are some issues with this: patches are counted twice in this spec (in `patches` and in the `package_hash`), the hashes of sources used to build are conflated with the `package.py` hash, and we don't actually include resources anywhere.

With this PR, I've expanded the package hash out in the `spec.json` body. Here is the "same" spec with the new fields:

```json
{
  "spec": {
    "_meta": {
      "version": 3
    },
    "nodes": [
      {
        "name": "zlib",
        "version": "1.2.12",
        ...
        "package_hash": "6kkliqdv67ucuvfpfdwaacy5bz6s6en4",
        "sources": [
          {
            "type": "archive",
            "sha256": "91844808532e5ce316b3c010929493c0244f3d37593afd6de04f71821d5136d9"
          }
        ],
        "patches": [
          "0d38234384870bfd34dfcb738a9083952656f0c766a0f5990b1893076b084b76"
        ],
        "hash": "ts3gkpltbgzr5y6nrfy6rzwbjmkscein"
      }
    ]
  }
}
```

Now:

* Patches and archive hashes are no longer included in the `package_hash`;
* Artifacts used in the build go in `sources`, and we tell you their checksum in the `spec.json`;
* `sources` will include resources for packages that have it;
* Patches are the same as before -- but only represented once; and
* The `package_hash` is a base32-encoded `sha1`, like other hashes in Spack, and it only
  tells you that the `package.py` changed.

The behavior of the DAG hash (which includes the `package_hash`) is basically the same as before, except now resources are included, and we can see differences in archives and resources directly in the `spec.json`

Note that we do not need to bump the spec meta version on this, as past versions of Spack can still read the new specs; they just will not notice the new fields (which is fine, since we currently do not do anything with them).

Among other things, this will more easily allow us to convert Spack specs to SBOM and track relevant security information (like `sha256`'s of archives). For example, we could do continuous scanning of a Spack installation based on these hashes, and if the `sha256`'s become associated with CVE's, we'll know we're affected.  (@vsoch FYI)

- [x] Add a method, `spec_attrs()` to `FetchStrategy` that can be used to describe a fetcher for a `spec.json`.
- [x] Simplify the way package_hash() is handled in Spack. Previously, it was handled as a special-case spec hash in `hash_types.py`, but it really doesn't belong there. Now, it's handled as part of `Spec._finalize_concretization()` and `hash_types.py` is much simpler.
- [x] Change `PackageBase.content_hash()` to `PackageBase.artifact_hashes()`, and include more information about artifacts in it.
- [x] Update package hash tests and make them check for artifact and resource hashes.

tgamblin committed:
specs: include source provenance in `spec.json` and package hash

We've included a package hash in Spack since #7193 for CI, and we started using it on
the spec in #28504. However, what goes into the package hash is a bit opaque. Here's
what `spec.json` looks like now:

```json
{
  "spec": {
    "_meta": {
      "version": 3
    },
    "nodes": [
      {
        "name": "zlib",
        "version": "1.2.12",
        ...
        "patches": [
          "0d38234384870bfd34dfcb738a9083952656f0c766a0f5990b1893076b084b76"
        ],
        "package_hash": "pthf7iophdyonixxeed7gyqiksopxeklzzjbxtjrw7nzlkcqleba====",
        "hash": "ke4alug7ypoxp37jb6namwlxssmws4kp"
      }
    ]
  }
}
```

The `package_hash` there is a hash of the concatenation of:

* A canonical hash of the `package.py` recipe, as implemented in #28156;
* `sha256`'s of patches applied to the spec; and
* Archive `sha256` sums of archives or commits/revisions of repos used to build the spec.

There are some issues with this: patches are counted twice in this spec (in `patches`
and in the `package_hash`), the hashes of sources used to build are conflated with the
`package.py` hash, and we don't actually include resources anywhere.

With this PR, I've expanded the package hash out in the `spec.json` body. Here is the
"same" spec with the new fields:

```json
{
  "spec": {
    "_meta": {
      "version": 3
    },
    "nodes": [
      {
        "name": "zlib",
        "version": "1.2.12",
        ...
        "package_hash": "6kkliqdv67ucuvfpfdwaacy5bz6s6en4",
        "sources": [
          {
            "type": "archive",
            "sha256": "91844808532e5ce316b3c010929493c0244f3d37593afd6de04f71821d5136d9"
          }
        ],
        "patches": [
          "0d38234384870bfd34dfcb738a9083952656f0c766a0f5990b1893076b084b76"
        ],
        "hash": "ts3gkpltbgzr5y6nrfy6rzwbjmkscein"
      }
    ]
  }
}
```

Now:

* Patches and archive hashes are no longer included in the `package_hash`;
* Artifacts used in the build go in `sources`, and we tell you their checksum in the `spec.json`;
* `sources` will include resources for packages that have it;
* Patches are the same as before -- but only represented once; and
* The `package_hash` is a base32-encoded `sha1`, like other hashes in Spack, and it only
  tells you that the `package.py` changed.

The behavior of the DAG hash (which includes the `package_hash`) is basically the same
as before, except now resources are included, and we can see differences in archives and
resources directly in the `spec.json`

Note that we do not need to bump the spec meta version on this, as past versions of
Spack can still read the new specs; they just will not notice the new fields (which is
fine, since we currently do not do anything with them).

Among other things, this will more easily allow us to convert Spack specs to SBOM and
track relevant security information (like `sha256`'s of archives). For example, we could
do continuous scanning of a Spack installation based on these hashes, and if the
`sha256`'s become associated with CVE's, we'll know we're affected.

- [x] Add a method, `spec_attrs()` to `FetchStrategy` that can be used to describe a
      fetcher for a `spec.json`.

- [x] Simplify the way package_hash() is handled in Spack. Previously, it was handled as
      a special-case spec hash in `hash_types.py`, but it really doesn't belong there.
      Now, it's handled as part of `Spec._finalize_concretization()` and `hash_types.py`
      is much simpler.

- [x] Change `PackageBase.content_hash()` to `PackageBase.artifact_hashes()`, and
      include more information about artifacts in it.

- [x] Update package hash tests and make them check for artifact and resource hashes.

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

iarspider commented:
@tgamblin since you will be working on package hashes, can you also try fixing #30720 ?

iarspider commented:
Quick question: will this fix the following problem - if Spack has a recipe for package, whose sources are taken from Git (or any other version control, I think), and I have a local repository with the same package but with different git URL, Spack will blindly use sources from Spack mirror?

tgamblin commented:
@spackbot run pipeline

scheibelp commented on var/spack/repos/builtin.mock/packages/hash-test4/package.py:
```diff
@@ -14,7 +14,7 @@ class HashTest4(Package):
     url = "http://www.hashtest1.org/downloads/hashtest4-1.1.tar.bz2"
 
-    version("1.1", "a" * 32)
+    version("1.1", sha256="aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
```
I think `"a" * 32` is clearer than the full string:

* We mandate the length of the hash
* Future readers may wonder why would someone go through the trouble of typing `"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"` when they could just type `"aaa"`
* Making it clear up front that the string is 32 characters is a good hint that there's a reason behind it

tgamblin replied:
ok all these are fixed

scheibelp commented on lib/spack/spack/hash_types.py:
```diff
@@ -5,28 +5,24 @@
+    The representation may only include certain dependency types, and it may optionally
+    include a canonicalized hash of the ``package.py`` for each node in the graph.
 
-    We currently use different hashes for different use cases."""
```
(question) are you removing this because it's no longer true, or because you don't think it's the right place to document this?

(we've discussed using a single hash everywhere for a while, but I don't think this PR handles that, so what I'm asking is if it already happened and this PR is just tidying up the docstring)

tgamblin replied:
It's redundant with the text above it, but if you think it's necessary, we could keep it.

scheibelp commented on lib/spack/spack/package_base.py:
```diff
@@ -1610,53 +1609,48 @@ def content_hash(self, content=None):
                 from_local_sources = env and env.is_develop(self.spec)
+
+                # don't bother setting a source id if none is available, but warn if
+                # it seems like there should be one.
```
If `source_id` is no longer going to be used in this function, then this comment should not refer to one.

tgamblin replied:
fixed

scheibelp commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -124,6 +125,20 @@ def __init__(self, **kwargs):
+            value = getattr(self, attr, None)
+            if value:
+                attrs[attr] = value
+        return attrs
```
It seems odd to embed configuration-YAML-awareness into the fetchers, what is this doing that `source_id` wasn't?

If this function is in fact necessary, I think it could use a better name and docstring (which I can suggest pending an answer to the above question).

tgamblin replied:
I tried to document this and remembered that I did not remove `source_id`.  See what you think of the new `source_provenance` function.

scheibelp commented on lib/spack/spack/package_base.py:
```diff
@@ -1597,7 +1596,7 @@ def all_patches(cls):
 
-    def content_hash(self, content=None):
+    def artifact_hashes(self, content=None):
         """Create a hash based on the artifacts and patches used to build this package.
```
I think the initial docstring should be updated:

* this is returning a list of hashes vs. a single hash combining the details described in the docstring.
* might want to explicitly state that resources are now included.

tgamblin replied:
fixed this.

scheibelp commented on lib/spack/spack/spec.py:
```diff
@@ -3030,12 +3044,16 @@ def _finalize_concretization(self):
 
-                # keep this check here to ensure package hash is saved
-                assert getattr(spec, ht.package_hash.attr)
+                # TODO: make artifact hashes a static method
```
I don't think this could ever be static: the artifact hashes include patches and resources (i.e. `_get_needed_resources` accesses `Package.spec` which is only available for instantiations).

tgamblin replied:
yes. the intent was to make it a *class* method *that takes a spec*. I clarified the comment.

tgamblin replied:
actually I just went ahead and made it a class method.

tgamblin replied:
Actually I just removed the comment b/c I don't think it's really needed. It is awkward to instantiate a package just for this but we can fix that later in a more broad way.

scheibelp commented on lib/spack/spack/spec.py:
```diff
@@ -3030,12 +3044,16 @@ def _finalize_concretization(self):
-                assert getattr(spec, ht.package_hash.attr)
+                # TODO: make artifact hashes a static method
+                artifact_hashes = pkg.artifact_hashes()
+                spec._package_hash = artifact_hashes.pop("package_hash")
```
It seems like it would be cleaner to call `package_hash()` directly here vs. depending on `artifact_hashes()` and then modifying the resulting dictionary: it seems like that's the direction that `artifact_hashes` is going (i.e. the name suggests we could exclude the contents of the package.py).

tgamblin replied:
Did you read the docstring for this method?

scheibelp replied:
My point was: however this was gotten in `pkg.artifact_hashes()`, get it directly here instead. I only see this called in one place in this PR, so I would have assumed calling

```
spack.util.package_hash(self)
```

directly (i.e. here, in this function) would make more sense than delegating it to `artifact_hashes` and then deconstructing it after the fact here.

scheibelp commented on lib/spack/spack/spec.py:
```diff
@@ -1917,23 +1921,24 @@ def to_node_dict(self, hash=ht.dag_hash):
+        # We use the attribute here instead of `self.package_hash()` because this should
+        # *always* be assigned at concretization time. We don't want to try to compute a
+        # package hash for concrete spec where a) the package might not exist, or b) the
+        # `dag_hash` didn't include the package hash when the spec was concretized.
```
(sanity check/question) this was already true for this function and you are documenting it here - did I get that right?

tgamblin replied:
the latter -- I tried to make the comment better.

scheibelp commented on lib/spack/spack/package_base.py:
```diff
@@ -1610,53 +1609,48 @@ def content_hash(self, content=None):
+        hashes = syaml.syaml_dict()
 
         # source artifacts/repositories
         # TODO: resources
```
This comment is no longer needed

tgamblin replied:
Thanks! fixed.

scheibelp commented on lib/spack/spack/spec.py:
```diff
@@ -1917,23 +1921,24 @@ def to_node_dict(self, hash=ht.dag_hash):
+                            return syaml.syaml_dict(sorted(source.items()))
+                        return source
+
+                    d[key] = [order(source) for source in source_list]
```
From looking through this PR it looks like the only thing that would ever need to be ordered is a dictionary. But FYI I had to spend some time cross-referencing to confirm that. If another editor ever adds a list of things, then this has the potential to become unordered (i.e. inconsistent hash).

You could add a consideration for lists in the locally-defined `order`, but that might be defeated by gaps in the type checking. Ideally if `_artifact_hashes` values are guaranteed to be iterables of comparable elements, then you could just call `sorted` without constructing a new type around it (i.e. don't create an `syaml_dict`: admittedly this interferes with the representation as listed in the PR description, which relates to point [1] of my review).

tgamblin replied:
I'm not sure I understand this comment. If the item is a list, it doesn't need to be deterministically ordered. It should already be.  We sort dictionaries but list order is up to the creator.

scheibelp replied:
Everything `artifact_hashes` calls (i.e. right now just `_get_needed_resources`) orders things, so this is fine for now. My worry is that this behavior is disconnected from the need: I had to dig a couple layers down to confirm this, vs. just seeing it enforced here where it really matters; that seems fragile to me. 

You may be arguing that we depend on this for other info we add in `to_node_dict`, but most of that is part of `_cmp_node` (which presumably tests order), and `_artifact_hashes` isn't.

scheibelp commented on lib/spack/spack/test/util/package_hash.py:
```diff
@@ -137,18 +135,38 @@ def test_content_hash_cannot_get_details_from_ast(mock_packages, config):
+    # the sources for these two packages will not be the same b/c their archive hashes differ
+    assert spec1.to_node_dict()["sources"] != spec2.to_node_dict()["sources"]
+
+    assert spec1.dag_hash() != spec2.dag_hash()
```
The specs have different package names, so these will always be unequal (i.e., I don't think this assert is confirming anything of value).

tgamblin replied:
ok I've removed this.

scheibelp reviewed (changes requested):
Point for consideration:

1. This is trending toward listing out more information individually in the node dict: what is the point of that? I don’t perceive a benefit other than that it becomes more-readable; IMO there are already mechanisms for getting a better interface to this information, in particular reading it in as a Spec object.

(I only have one point to make here but I numbered it because I reference this in one of the PR comments)

DarylGrunau commented:
Just pinging here to see what the status/blockers are for this MR to be merged.  Having the capability to produce an SBOM is becoming highly desirable at LANL (perhaps there's another MR that does this?).

vsoch commented:
I had done https://github.com/spack/spack-sbom and a follow up PR here to add a command in 2021 (since closed) but you probably want the changes in first that @tgamblin thinks are important.

tgamblin commented:
@scheibelp: RE:

> This is trending toward listing out more information individually in the node dict:

Yes. 

> what is the point of that? I don’t perceive a benefit other than that it becomes more-readable; IMO there are already mechanisms for getting a better interface to this information, in particular reading it in as a Spec object.

There are not reliable interfaces for retrieving this from an old installation.  `package.py` files may drift; we need an authoritative place for this stuff. The Spec is where authoritative provenance goes.

I put some motivation in the description, but here it is for posterity:

> We want this information to be available *from the concrete spec* so that it can be retrieved as provenance *without* referring to a `package.py` file.  Why?
> 
> 1. We want to be able to look up authoritative checksums, versions, etc. and pass them through scanners long *after* an installation takes place.
> 2. Right now you can look at the version on a `Spec`, then look up the checksum for a version from the `package.py`, but that's not reliable in the long term.  We may remove versions from `package.py` (in fact, we might want to as they become less relevant), and checksums for particular versions/releases may change over time.
> 3. Only the version is stored explicitly on the spec; that's not sufficient to know *exactly* what hashes a spec was built with.
> 
> Adding hashes explicitly gets around this and gets us closer to having all the information we need for a complete SBOM.

tgamblin committed:
extra debug statements

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

greenc-FNAL commented:
Does the [launch](https://thenewstack.io/protobom-paving-the-path-for-sbom-adoption/) of [Protobom](https://openssf.org/projects/protobom/) affect this PR in any way? Alternatively, should there be read-in converters for `spec.yaml` and `spack.lock` files in [Protobom](http://github.com/bom-squad/protobom) itself?

vsoch commented:
oh man, protobom! I love protocol buffers so this is :pinched_fingers:

tgamblin committed:
remove TODO: resources

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
update docstring

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
don't refer to source id

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
abbreviate hashes in mock packages

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
clarify comment on use of _package_hash

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
remove useless check

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
remove comment

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
avoid circular import of `spack.patch`

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
fix string in test

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
installer: show tracebacks from builds to debug failed installs

`installer.py` currently swallows the traceback and preserves only
the error messaege if a build process fails.

- [x] preserve exceptions from failed build processes
- [x] print a full traceback for each one when running with `spack -d`

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
don't remove patches variant

Patches variant needs to stay on the spec even if we also have resources.

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
bugfix: submodule callbacks need to be resolved in specs

The submodules argument for git versions supports callbacks (which are used on some key
packages to conditionally fetch certain submodules). Callbacks can't be serialized to
JSON, so we need to ensure that these things are resolved (by calling them) at
concretization time.

- [x] ensure that submodule callbacks are called when creating spec JSON
- [x] add tests that try serializing all types of git versions

scheibelp commented on lib/spack/spack/test/cmd/install.py:
```diff
@@ -812,28 +821,92 @@ def test_install_no_add_in_env(tmpdir, mock_fetch, install_mockery, mutable_mock
+        print(initial_libelf == final_libelf)
+        print(id(initial_libelf), id(final_libelf))
+        print(initial_libelf in initial_concrete_specs)
+        print(initial_libelf in post_install_concrete_specs)
```
This looks like leftover debugging code (or some of it at least seems too detailed - like printing out the `_cmp_iter` results with no accompanying context - and I thought `print` would dump to the terminal in the middle of the test vs. `tty.debug`).

tgamblin replied:
> I thought print would dump to the terminal in the middle of the test vs. tty.debug).

It won't -- on success this is all hidden. On failure it's shown. But it's too much anyway so I got rid of it.

scheibelp commented on lib/spack/spack/test/spec_yaml.py:
```diff
@@ -96,6 +96,22 @@ def test_invalid_json_spec(invalid_json, error_message):
+        "git-url-top-level@2.2",
+        "git-url-top-level@2.1",
+        "git-url-top-level@2.0",
+        "git-url-top-level@2.3",
```
* `git-url-top-level@2.3` appears twice in this list
* AFAICT, 2.3 and 2.2 are the same thing, and likewise for 2.1 and 2.0
* The 2.x versions are URL-based, so IMO the leading comment is confusing: `all types of git versions`
* `check_x_round_trip` uses `Spec.eq_dag`: since none of the added functionality in this PR appears in `_cmp_node`, I'm not sure this is the test to use (i.e. I think it would continue to pass even if someone deleted one "side" of `to_node_dict`/`from_node_dict`)

tgamblin replied:
I addressed all of this but the last one -- I am not sure what you mean with the last point. I think it's still important to make sure these can be round-tripped.

scheibelp commented on lib/spack/spack/package_base.py:
```diff
@@ -1667,66 +1666,75 @@ def all_patches(cls):
-            hash_content.extend(
-                ":".join((p.sha256, str(p.level))).encode("utf-8") for p in self.spec.patches
-            )
+            hashes["patches"] = [p.sha256 for p in self.spec.patches]
```
Applying a patch at a different level is a significant difference: why remove this?

(no change needed) Also, it appears you dropped all calls to `.encode(utf-8)`: that makes sense given the removal of computations using `hashlib` (handling with `spack.util.hash.*` instead, which deals with encoding).

tgamblin replied:
good point -- added this back.

tgamblin replied:
Ok, adding this back breaks spec reading, which is something I had wanted to avoid.  Apparently:

1. We encode `patches` as a variant.
2. We encode `patches` at the top level of the spec dictionary as a list of hashes, and it needs to stay there to be read properly by old versions of Spack.  I *could* bump the spec version, but was hoping not to.  Trying to decide whether we should.
3. We should preserve patch level, but I don't think that it can go in either of the other two places. The place to put it is *probably* in the sources list, along with other artifacts. I don't really like this b/c now `patches` will appear in 3 places on the spec, but maybe the others can be deprecated.  Maybe this is a good enough reason to bump the spec version, or at least something we should do when we next bump it.

Still deciding what to do here.

scheibelp commented on lib/spack/spack/spec.py:
```diff
@@ -1440,6 +1440,12 @@ def __init__(
         for h in ht.hashes:
             setattr(self, h.attr, None)
 
+        # dictionary of source artifact hashes, set at concretization time
```
I think this comment is copied from next attr (and does not accurately describe what `_package_hash` is).

tgamblin replied:
fixed this one

scheibelp reviewed (commented):
> package.py files may drift; we need an authoritative place for this stuff. 

If they drift, the package_hash should catch that

> The Spec is where authoritative provenance goes.

While I definitely agree, I worry about storing the same provenance in redundant ways in the ~spec~ node dictionary representation, purely for readability for end users (for a storage format that, AFAIK, was not intended for API purposes other than to be a stable form of storage for our own reconstruction).

One thing I want to check: is this intended to precisely identify the commit that we end up using for source control repo branch tips (e.g. develop in a git repo)? I don't see that right now in `spec_attrs`, but if that isn't the aim, then I'm not sure why this implementation retrieves info from the fetcher, when all the provenance should be on the version declaration. This is related to https://github.com/spack/spack/pull/32312/files#r1041676449

tgamblin committed:
remove straggling uses of hash_types.package_hash

- [x] don't allow calling `spec.package_hash()` on abstract specs
- [x] get rid of last few uses of `ht.package_hash`, which was removed.

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
remove superfluous debug lines in test

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
source_id -> source_provenance

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
remove redundant list element

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
clean up test

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
add back patch level

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>
